### PR TITLE
[Runtimes] General fixes and adding FunciontReference

### PR DESCRIPTION
**Pull request by yaronha**
* various function nuclio runtime fixes 
* fix debug env var
* add py requierments option to code to func and fn.with_requierments()
* add function ref object
* move fn.with_code() to runtime base

### CODE REVIEW

File: mlrun/runtimes/funcref.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
+# Copyright 2018 Iguazio
```
Comment (Hedingber) [inline mlrun/runtimes/funcref.py:1]: Let's call the file `function_reference`

File: mlrun/runtimes/funcref.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+# limitations under the License.
+
+import os
+import mlrun
+
+from ..model import ModelObj
+from ..utils import generate_object_uri
+
+
+class FunctionRef(ModelObj):
```
Comment (Hedingber) [inline mlrun/runtimes/funcref.py:22]:
```suggestion
class FunctionReference(ModelObj):
```

File: mlrun/runtimes/funcref.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    def __init__(
+        self,
+        url=None,
+        image=None,
+        requirements=None,
+        code=None,
+        spec=None,
+        kind=None,
+        name=None,
```
Comment (Hedingber) [inline mlrun/runtimes/funcref.py:33]: Can you elaborate why would a function reference have anything else than... the reference which is the URL, what's the motivation ?

**Review by Hedingber — CHANGES_REQUESTED** [review]

File: mlrun/runtimes/funcref.py (hunk 1 above)
Comment (yaronha) [reply mlrun/runtimes/funcref.py:1]: ok

File: mlrun/runtimes/funcref.py (hunk 2 above)
Comment (yaronha) [reply mlrun/runtimes/funcref.py:22]: ok

File: mlrun/runtimes/funcref.py (hunk 3 above)
Comment (yaronha) [reply mlrun/runtimes/funcref.py:33]: this is more of a template, not a full func, and user want to be able to specify/load from a simple py file/notebook, not forced to create a function spec, we do the same in project (we can use this object under project as well), the spec part allows us to place full func, or part that we want to add (env/volumes/gpu). so we need a way to specify the image/kind and requirements.

File: mlrun/runtimes/funcref.py (hunk 1 above)
Comment (yaronha) [reply mlrun/runtimes/funcref.py]: fixed

File: mlrun/runtimes/funcref.py (hunk 2 above)
Comment (yaronha) [reply mlrun/runtimes/funcref.py]: fixed

**Review by Hedingber — APPROVED** [review]

**Pull request merged**